server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -580,20 +597,93 @@ public synchronized void close() throws IOException {
+        CompletableFuture<Void> allStoppers = CompletableFuture.allOf(futures.values().toArray(new CompletableFuture[stoppers.size()]));
+
+        TimeValue maxTimeout = MAXIMUM_SHUTDOWN_TIMEOUT_SETTING.get(this.settings());
         try {
-            stopper.get();
-        } catch (Exception e) {
-            logger.warn("unexpected exception while waiting for http server to close", e);
+            if (maxTimeout.millis() == 0) {
+                FutureUtils.get(allStoppers);
+            } else {
+                FutureUtils.get(allStoppers, maxTimeout.millis(), TimeUnit.MILLISECONDS);
+            }
+
+        } catch (ElasticsearchTimeoutException t) {
+            var unfinishedTasks = futures.entrySet()
+                .stream()
+                .filter(entry -> entry.getValue().isDone() == false)
+                .map(Map.Entry::getKey)
+                .toList();
+            logger.info("timed out while waiting for graceful shutdown tasks: " + unfinishedTasks);
+        }
+    }
+
+    private void awaitSearchTasksComplete() {
+        TaskManager taskManager = injector.getInstance(TransportService.class).getTaskManager();
+        TimeValue asyncSearchTimeout = SHUTDOWN_SEARCH_TIMEOUT_SETTING.get(this.settings());
+        long millisWaited = 0;
+        while (true) {
+            long searchTasksRemaining = taskManager.getTasks()
+                .values()
+                .stream()
+                .filter(task -> TransportSearchAction.TYPE.name().equals(task.getAction()))
+                .count();
+            if (searchTasksRemaining == 0) {
+                logger.debug("all search tasks complete");
+                return;
+            } else {
+                // Let the system work on those searches for a while. We're on a dedicated thread to manage app shutdown, so we
+                // literally just want to wait and not take up resources on this thread for now. Poll period chosen to allow short
+                // response times, but checking the tasks list is relatively expensive, and we don't want to waste CPU time we could
+                // be spending on finishing those searches.
+                final TimeValue pollPeriod = TimeValue.timeValueMillis(500);
+                millisWaited += pollPeriod.millis();
+                if (millisWaited >= asyncSearchTimeout.millis()) {
+                    logger.warn(
+                        format(
+                            "timed out after waiting [%s] for [%d] search tasks to finish",
+                            asyncSearchTimeout.toString(),
+                            searchTasksRemaining
+                        )
+                    );
+                    return;
+                }
+                logger.debug(format("waiting for [%s] search tasks to finish, next poll in [%s]", searchTasksRemaining, pollPeriod));
+                try {
+                    Thread.sleep(Math.min(pollPeriod.millis(), (asyncSearchTimeout.millis() - millisWaited)));
```

Comment:
I think it's ok to simplify this to just the poll period. If we timeout in the middle of a poll that's ok.

Reply:
It actually will time out a maximum of one poll period too soon, which is probably okay since on the scale of these timeouts I'm not sure 500ms is going to matter in terms of letting a search complete or not, and that's better than it going too long and the central handler timing out since we get more information if we time out here. In any event I've simplified it.